allennlp/data/iterators/lazy_basic_iterator.py

```diff
@@ -0,0 +1,96 @@
+from typing import Iterator, Dict, List, Iterable
+import itertools
+import logging
+
+from overrides import overrides
+
+from allennlp.common import Params
+from allennlp.data.iterators.data_iterator import DataIterator
+from allennlp.data.dataset import LazyDataset, Dataset
+from allennlp.data.instance import Instance
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+@DataIterator.register("lazy-basic")
+class LazyBasicIterator(DataIterator[LazyDataset]):
+    """
+    A very basic lazy iterator, which takes a dataset and yields fixed size batches.
+    Each batch is padded to the maximum length within that batch.
+
+    Parameters
+    ----------
+    batch_size : int, optional, (default = 32)
+        The size of each batch of instances yielded when calling the iterator.
+    instances_per_epoch: int, optional, (default = None)
+        If this parameter is specified, each epoch generated by ``LazyIterator.__call__``
+        contains at most this many instances. A "cursor" is stored for each Dataset,
+        so that each epoch picks up where the previous one left off.
+
+        If the value is ``None``, then each epoch is just a full pass through the Dataset.
+    """
+    def __init__(self, batch_size: int = 32, instances_per_epoch: int = None) -> None:
+        self._batch_size = batch_size
+        self._instances_per_epoch = instances_per_epoch
+
+        # As you might use this iterator with multiple datasets,
+        # you need to store a cursor for each one.
+        self._cursors: Dict[LazyDataset, Iterator[Instance]] = {}
+
+    @overrides
+    def get_num_batches(self, _: LazyDataset) -> int:
+        # pylint: disable=no-self-use
+        # TODO(joelgrus): Figure out the right way to handle this when `instances_per_epoch`
+        #   is not provided. This is definitely not right, but currently this is only used for tqdm,
+        #   and the only consequence is that the tqdm for iterating over an epoch
+        #   will show as a count rather than as a progress bar.
+        return self._instances_per_epoch // self._batch_size if self._instances_per_epoch else 1
```

Comment:
It seems to me that we just need to decide what the default is: 1 instance per epoch or infinity instances per epoch? Then we don't have to worry about num_instances_per_epoch being none and we just return the default here.

Personally, I think having the default be "infinity" epochs might do the right thing here, but I'm not familiar enough with how `get_num_batches` is used yet to say if it'd create bugs.